### ISSUE TITLE
Improve GenerateTransactionId() function performance

### ISSUE THREAD

**Issue by jniles** [third-culture-software/bhima#3002]
@lomamech points out that our SQL function [`GenerateTransactionId()`](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/server/models/functions.sql#L73) function is very inefficient.  We could improve this easily enough by rewriting it to store the latest transaction ID for each project in a small table.  The table would look like this:

| **project_id** | **transaction_id** |
|:--:|:--:|
| 1 | 254 |
| 2 | 576 |

The logic is that that every time a transaction ID is successfully used/generated, the `transaction_id` column is updated.  This means that instead of a full scan of the `posting_journal` or `general_ledger`, it will just pluck the latest version from the table.

As much as possible, this call should not change the signature of the `GenerateTransactionId()` function.

**Comment by jmcameron** [comment]
Note that this proposal has the potential of being brittle since the table may be updated and used in different places, possibly by multiple sections of code.  Therefore it will be extra important to have good testing of this feature to make sure the table us updated EVERY time a transaction ID is generated or used anywhere in the code.

**Comment by sfount** [comment]
@lomamech @jniles @jmcameron improving something that is called so frequently will make a huge difference on low end hardware. 

**In brief:** With an indexed integer representation of the transaction ID we could improve this query by ~200% (0.157624s vs. 0.000883s as briefly tested)

I reviewed the query this morning  (`SHOW PROFILE;`)  and it looks like the element of the profile that is most time consuming is interestingly **'Sending data'**. My understanding here is that MySQL has to actually load (very temporarily) every single `trans_id` into memory to then perform the `SUBSTR`, `FLOOR`, and `MAX`. This could easily be 200,000+ strings.

The inefficiency is most likely in transporting these strings around rather than any joins or index issues. 

One solution that addresses the multiple sections of code consistency pointed out by @jmcameron and the performance would be to store the integer part of the ID in both the `posting_journal` and the `general_ledger` as an indexed column. This would allow us to write this query with a more succinct limit order.

```SQL 
ORDER BY trans_id_integer LIMIT 1;
```
In theory this should improve performance but guard against allowing non-transactional updates from different areas of the code. 

A quick analysis on ~200,000 records and averaged across multiple queries:

Original query (copied from repo):

| Roundtrip Query Duration | Query Profile Duration | Sending Data Duration  |
| ------------- | ------------- | ----- |
| 0.516s    | 0.157624s | 0.154132 |

The key thing to note here is that 'Sending Data' is by far the majority of the query time. 

![old_query](https://user-images.githubusercontent.com/2844572/44848672-9a8b1480-ac4f-11e8-9651-52239de6f670.png)

Using `EXPLAIN` we can see that each table is having to access every one of its **rows** in order to produce the result, crucially *this complexity grows with the size of the table*.

By adding a new indexed column to these tables we can do a new query that attempts to visit far less rows in order to get the same result, for example: 

```sql 
SELECT trans_id_integer_part 
FROM posting_journal
WHERE project_id = 1
ORDER BY trans_id_integer_part 
LIMIT 1;
```
Briefly running the equivalent query using this method (including UNION etc.)

| Roundtrip Query Duration | Query Profile Duration | Sending Data Duration  |
| ------------- | ------------- | ----- |
| 0.000s    | 0.000883s | 0.000017 |

This has reduced the query complexity *at the cost of increased storage requirements for the new index*.

![new_query](https://user-images.githubusercontent.com/2844572/44848854-2c931d00-ac50-11e8-95f4-cde9b66d9202.png)

This query only accesses one row from these giant tables, this is performed in O(1).

Implementing this solution would require upgrading current production installations of the software so it should be considered carefully. 

* Database schema changed to add an additional column and index to the `posting_journal` and `general_ledger`
* Upgrade script for current databases to add the new column and index, as well as populate the new column 
* Update all scripts writing to the `posting_journal` and `general_ledger` to also write the integer (reference) part of the transaction ID 
* Update the `GenerateTransactionID` function to take advantage of the new indexed column

It should be noted that these test have been run on a desktop multi-core i7 processor and the results should have a more profound effect on lower end machines.

The SQL used to modify the database for this test was: 
```sql
alter table posting_journal ADD trans_id_reference_part INT;
alter table general_ledger ADD trans_id_reference_part INT;

update posting_journal set trans_id_reference_part = SUBSTR(trans_id, 4);
update general_ledger set trans_id_reference_part = SUBSTR(trans_id, 4);

create index reference_part on posting_journal (trans_id_reference_part);
create index reference_part on general_ledger (trans_id_reference_part);
```

**Comment by jniles** [comment]
@sfount, great analysis!  This gives us some good data to make our determination.

Honestly, putting the integer part in the posting_journal and general_ledger is probably the best way to go.  It gives a clear history of what happened if something goes wrong and doesn't require any extra triggers.  If you want to work on a PR for this, it would be greatly appreciated 👍

**Comment by jniles** [comment]
Closed in #3110
